### DISCUSSION TITLE
Typechecking with Fields in 3.10

### DISCUSSION THREAD

**Discussion by Atticus1806** [agronholm/typeguard#381] [Q&A]
### Things to check first

- [X] I have searched the existing issues and didn't find my bug already reported there

- [X] I have checked that my bug is still present in the latest release

### Typeguard version

4.0.0

### Python version

3.10

### What happened?

For our work we want to implement type-checked dataclasses. We currently did this via this code:
```
@dataclass
class DataclassDummy:
    def _validate_types(self) -> None:
        for field in fields(type(self)):
            try:
                    typeguard.check_type(getattr(self, field.name), field.type)
            except typeguard.TypeCheckError as exc:
                raise typeguard.TypeCheckError(f'In field "{field.name}" of "{type(self)}": {exc}')

    def __post_init__(self) -> None:
        self._validate_types()
```

This does not work with python 3.10 since `field.type` returns a string. Sometimes `typeguard` catches this and "ignores" the type with a warning, sometimes it crashes. Currently we have a hacky solution to this problem but want to ask if there is some cleaner way to solve it: 
```
    def _validate_types(self) -> None:
        for field in fields(type(self)):
            try:
                if isinstance(field.type, str):
                    from typeguard._functions import check_type_internal
                    from typeguard._memo import TypeCheckMemo
                    from typing import ForwardRef
                    import sys
                    cls_globals = vars(sys.modules[self.__class__.__module__])
                    memo = TypeCheckMemo(globals=cls_globals, locals=cls_globals)
                    t = ForwardRef(field.type)
                    check_type_internal(getattr(self, field.name), t, memo)
                else:
                    typeguard.check_type(getattr(self, field.name), field.type)
            except typeguard.TypeCheckError as exc:
                raise typeguard.TypeCheckError(f'In field "{field.name}" of "{type(self)}": {exc}')
```
This works in our setup and produces (at least for our tests) correct behavior. Is there something similar known? Is there a better workaround?

### How can we reproduce the bug?

See above.

**Comment by agronholm** [comment]
You could resolve the type references to actual types before handing them over to `check_type()`. Cache the references too because resolving the references every time is a bit resource intensive.
